src/api/classify_element.rs

```diff
@@ -0,0 +1,349 @@
+        assert!( branches <= 1000, format!("Classify Element branches: {} > 1000", branches));
+        assert!( queue_capacity <= 1000, format!("Classify Element queue_capacity: {} > 1000", queue_capacity));
+
+        let mut to_providers: Vec<Sender<Option<E::Packet>>> = Vec::new();
+        let mut await_providers: Vec<Sender<task::Task>> = Vec::new();
+        let mut wake_providers: Vec<Receiver<task::Task>> = Vec::new();
+        let mut providers: Vec<ClassifyElementProvider<E>> = Vec::new();
+
+        for _ in 0..branches {
+            let (to_provider, from_consumer) = crossbeam_channel::bounded::<Option<E::Packet>>(queue_capacity);
+            let (await_consumer, wake_provider) = crossbeam_channel::bounded::<task::Task>(1);
+            let (await_provider, wake_consumer) = crossbeam_channel::bounded::<task::Task>(1);
+
+            let provider = ClassifyElementProvider::new(from_consumer, await_consumer, wake_consumer);
+
+            to_providers.push(to_provider);
+            await_providers.push(await_provider);
+            wake_providers.push(wake_provider);
+            providers.push(provider);
+        }
+
+        ClassifyElementLink {
+            consumer: ClassifyElementConsumer::new(input_stream, to_providers, element, await_providers, wake_providers),
+            providers
+        }
+    }
+}
+
+pub struct ClassifyElementConsumer<E: ClassifyElement> {
+    input_stream: ElementStream<E::Packet>,
+    to_providers: Vec<Sender<Option<E::Packet>>>,
+    element: E,
+    await_providers: Vec<Sender<task::Task>>,
+    wake_providers: Vec<Receiver<task::Task>>
+}
+
+impl<E: ClassifyElement> ClassifyElementConsumer<E> {
+    fn new(
+        input_stream: ElementStream<E::Packet>,
+        to_providers: Vec<Sender<Option<E::Packet>>>,
+        element: E,
+        await_providers: Vec<Sender<task::Task>>,
+        wake_providers: Vec<Receiver<task::Task>>
+    ) -> Self {
+        ClassifyElementConsumer {
+            input_stream,
+            to_providers,
+            element,
+            await_providers,
+            wake_providers
+        }
+    }
+}
+
+impl<E: ClassifyElement> Drop for ClassifyElementConsumer<E> {
+    fn drop(&mut self) {
+        //TODO: do this with a closure or something, this could be a one-liner
+        for to_provider in self.to_providers.iter() {
+            if let Err(err) = to_provider.try_send(None) {
+                panic!("Consumer: Drop: try_send to provider, fail?: {:?}", err);
+            }
+        }
+
+        for wake_provider in self.wake_providers.iter() {
+            if let Ok(task) = wake_provider.try_recv() {
+                task.notify();
+            }
+        }
+    }
+}
+
+impl<E: ClassifyElement> Future for ClassifyElementConsumer<E> {
+    type Item = ();
+    type Error = ();
+
+    /// Same logic as AsyncElementConsumer, except if any of the channels are full we
```

Comment:
We can do this in a separate PR, but if we're sharing similar logic with `AsyncElementConsumer`, we should be thinking about what parts are similar and if we can compose any `AsyncElement` components here to reduce the complexity.

Reply:
https://github.com/route-rs/route-rs/issues/29 Created issue.